external-crates/move/crates/move-compiler/src/sui_mode/linters/custom_rules/shift_overflow.rs

```diff
@@ -0,0 +1,121 @@
+//! Detect potential overflow scenarios where the number of bits being shifted exceeds the bit width of
+//! the variable being shifted, which could lead to unintended behavior or loss of data. If such a
+//! potential overflow is detected, a warning is generated to alert the developer.
+use crate::{
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        WarningFilters,
+    },
+    expansion::ast::Value_,
+    naming::ast::{BuiltinTypeName_, TypeName_, Type_},
+    parser::ast::BinOp_,
+    shared::{program_info::TypingProgramInfo, CompilationEnv},
+    sui_mode::linters::{LinterDiagCategory, LINTER_DEFAULT_DIAG_CODE, LINT_WARNING_PREFIX},
+    typing::{
+        ast::{self as T, UnannotatedExp_},
+        visitor::{TypingVisitorConstructor, TypingVisitorContext},
+    },
+};
+use move_ir_types::location::Loc;
+use std::str::FromStr;
+
+const SHIFT_OPERATION_OVERFLOW_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::ShiftOperationOverflow as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "Potential overflow detected. The number of bits being shifted exceeds the bit width of the variable being shifted.",
+);
+
+pub struct ShiftOperationOverflowVisitor;
+
+pub struct Context<'a> {
+    env: &'a mut CompilationEnv,
+}
+
+impl TypingVisitorConstructor for ShiftOperationOverflowVisitor {
+    type Context<'a> = Context<'a>;
+
+    fn context<'a>(
+        env: &'a mut CompilationEnv,
+        _program_info: &'a TypingProgramInfo,
+        _program: &T::Program_,
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
+        if let UnannotatedExp_::BinopExp(lhs, op, _, rhs) = &exp.exp.value {
+            if matches!(op.value, BinOp_::Shl | BinOp_::Shr) {
+                if let Some(bit_width) = get_bit_width(&lhs.ty.value) {
+                    if let Some(shift_amount) = get_shift_amount(&rhs.exp.value) {
```

Comment:
and_then should help here. Or some let-else

Reply:
Checkout new PR for this lint #16820